src/v/v8_engine/instance.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/core/temporary_buffer.hh>
+
+#include <v8.h>
+
+namespace v8_engine {
+
+class instance_exception final : public std::exception {
+public:
+    explicit instance_exception(ss::sstring msg)
+      : _msg(std::move(msg)) {}
+
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+// This class implemented container for run js code inside it.
+// It stores only one function from js code and can run it.
+// v8::Isolate are used like container for js env.
+// It can help isolate different functions from each other
+// and set up constraints for them separately.
+class instance {
```

Comment:
Maybe we should change the name? This makes me think its an instance of the v8 engine since the fully qualified name is `v8_engine::instance`. I think maybe `wasm_function` or just `function` is appropriate

Reply:
this isnt for a 'function' but per core. a v8::isolate should be able to handle multiple scripts.

Reply:
Rob is right
We use one isolate for one function. It allows us set memory limit.
If we use one isolate for one core we can not run functions parallel in this isolate in executor (it will be a thread pool). Because for run a function we need to lock isolate.

Reply:
why have one isolate per function. I thought Candy explored this on her talk no? what's the reasoning behind this?

Why can't it be a `a list of compiled_script` and we swap them in and out at runtime. 

Ok for this to come at a later date, but makes sense to me to allow for the isolate to swap functions in and out.

Reply:
cc: @dotnwat  - you should chime in here

Reply:
@VadimPlh  btw; this is ok if it comes in a future PR. 

Can you actually put together an RFC for this feature. I just thought this was infrastructure PR, not a committed design yet.

Reply:
As I remembered, Candy did not say nothing about it.
If we use a lot functions inside one isolate we can not control how much memory each function used. In shredder they did not do work about script restrictions.

Reply:
The main idea is
1) use  `ConfigureDefaultsFromHeapSize` for each isolate(= function)
2) run different isolate in parallel in executor

Reply:
makes sense. i think the RFC would be good to capture these thoughts from you, so we can all learn. 

what's the memory use per function. at some point we'd need to limit the memory use per function/script. the reason that's nice is we can set a global limit of say 1GB or 500MB per core, up front and reserved, so regardless of how many scripts, they would all have a limit of the memory they use.

Reply:
Also, you can set different timeout for different function. If we use different isolates it can not affect for other function.

If we use only one, It affect for other function call

Reply:
@senior7515 Yes, I need push some docs. But first of all I want to finish with v8 build.

Reply:
makes sense, ok to defer to later pull requests. thanks for sharing.

Reply:
@senior7515 
Looks like Cloudflare use one isolate per one function
https://developers.cloudflare.com/workers/learning/how-workers-works
https://developers.cloudflare.com/workers/learning/security-model

Reply:
Interesting!!!

Reply:
Maybe script? @graphcareful